File: backend/apps/team_management/schemas.py
from datetime import date
from decimal import Decimal
from typing import Literal
from uuid import UUID

from ninja import Schema
from pydantic import Field


TeamTypeValue = Literal[
    "executive",
    "management",
    "engineering",
    "design",
    "marketing",
    "sales",
    "finance",
    "operations",
    "support",
    "hr",
    "legal",
    "project",
    "custom",
]

EmploymentStatusValue = Literal[
    "active",
    "on_leave",
    "suspended",
    "resigned",
    "terminated",
    "contract_ended",
    "inactive",
]

EngagementTypeValue = Literal[
    "full_time",
    "part_time",
    "contract",
    "intern",
    "consultant",
    "volunteer",
]

WorkLocationTypeValue = Literal[
    "onsite",
    "remote",
    "hybrid",
]

ExpertiseLevelValue = Literal[
    "awareness",
    "working",
    "proficient",
    "expert",
    "lead",
]


class TeamCreateSchema(Schema):
    name: str
    slug: str
    team_type: TeamTypeValue = "custom"
    description: str = ""
    parent_id: UUID | None = None
    manager_id: UUID | None = None
    is_active: bool = True
    is_public: bool = False
    sort_order: int = 0
    metadata: dict = Field(default_factory=dict)


class TeamUpdateSchema(TeamCreateSchema):
    pass


class TeamMembershipInputSchema(Schema):
    team_id: UUID
    role_title: str = ""
    is_primary: bool = False
    is_active: bool = True
    joined_at: date | None = None
    left_at: date | None = None
    sort_order: int = 0


class TeamMemberServiceInputSchema(Schema):
    service_id: UUID
    expertise_level: ExpertiseLevelValue = "proficient"
    years_of_experience: Decimal | None = None
    is_primary: bool = False
    is_public: bool = True
    sort_order: int = 0


class TeamMemberCreateSchema(Schema):
    user_id: UUID | None = None
    employee_code: str
    first_name: str
    last_name: str = ""
    preferred_name: str = ""
    job_title: str
    professional_title: str = ""
    email: str = ""
    phone: str = ""
    public_email: str = ""
    public_phone: str = ""
    profile_image_id: UUID | None = None
    bio: str = ""
    short_bio: str = ""
    qualifications: str = ""
    years_of_experience: int | None = None
    engagement_type: EngagementTypeValue = "full_time"
    employment_status: EmploymentStatusValue = "active"
    work_location_type: WorkLocationTypeValue = "onsite"
    office_location: str = ""
    country: str = ""
    timezone_name: str = "Asia/Colombo"
    joined_at: date | None = None
    employment_ended_at: date | None = None
    reports_to_id: UUID | None = None
    linkedin_url: str = ""
    github_url: str = ""
    portfolio_url: str = ""
    website_url: str = ""
    is_leadership: bool = False
    is_public: bool = False
    is_featured: bool = False
    sort_order: int = 0
    metadata: dict = Field(default_factory=dict)
    memberships: list[
        TeamMembershipInputSchema
    ] = Field(default_factory=list)
    services: list[
        TeamMemberServiceInputSchema
    ] = Field(default_factory=list)


class TeamMemberUpdateSchema(TeamMemberCreateSchema):
    pass


class TeamMembershipSchema(Schema):
    id: UUID
    team_id: UUID
    team_name: str
    role_title: str
    is_primary: bool
    is_active: bool
    joined_at: date
    left_at: date | None
    sort_order: int


class TeamMemberServiceSchema(Schema):
    id: UUID
    service_id: UUID
    service_title: str
    expertise_level: str
    years_of_experience: Decimal | None
    is_primary: bool
    is_public: bool
    sort_order: int


class TeamMemberSchema(Schema):
    id: UUID
    user_id: UUID | None
    employee_code: str
    first_name: str
    last_name: str
    full_name: str
    preferred_name: str
    display_name: str
    job_title: str
    professional_title: str
    email: str
    phone: str
    public_email: str
    public_phone: str
    profile_image_id: UUID | None
    bio: str
    short_bio: str
    qualifications: str
    years_of_experience: int | None
    engagement_type: str
    employment_status: str
    work_location_type: str
    office_location: str
    country: str
    timezone_name: str
    joined_at: date | None
    employment_ended_at: date | None
    reports_to_id: UUID | None
    reports_to_name: str | None
    linkedin_url: str
    github_url: str
    portfolio_url: str
    website_url: str
    is_leadership: bool
    is_public: bool
    is_featured: bool
    is_current: bool
    sort_order: int
    metadata: dict
    memberships: list[TeamMembershipSchema]
    services: list[TeamMemberServiceSchema]


class TeamSchema(Schema):
    id: UUID
    name: str
    slug: str
    team_type: str
    description: str
    parent_id: UUID | None
    parent_name: str | None
    manager_id: UUID | None
    manager_name: str | None
    is_active: bool
    is_public: bool
    sort_order: int
    metadata: dict
    member_count: int


class PublicTeamMemberSchema(Schema):
    id: UUID
    display_name: str
    job_title: str
    professional_title: str
    public_email: str
    public_phone: str
    profile_image_id: UUID | None
    short_bio: str
    bio: str
    qualifications: str
    years_of_experience: int | None
    country: str
    linkedin_url: str
    github_url: str
    portfolio_url: str
    website_url: str
    is_leadership: bool
    is_featured: bool
    services: list[TeamMemberServiceSchema]


class PublicTeamSchema(Schema):
    id: UUID
    name: str
    slug: str
    team_type: str
    description: str
    members: list[PublicTeamMemberSchema]



class TeamMemberStatusUpdateSchema(Schema):
    employment_status: EmploymentStatusValue
    employment_ended_at: date | None = None


class TeamMemberReportingLineSchema(Schema):
    reports_to_id: UUID | None = None


class TeamManagerUpdateSchema(Schema):
    manager_id: UUID | None = None


class TeamManagementDashboardSchema(Schema):
    total_teams: int
    active_teams: int
    public_teams: int
    total_members: int
    active_members: int
    members_on_leave: int
    inactive_members: int
    public_members: int
    featured_members: int
    leadership_members: int
    members_without_primary_team: int
    members_without_manager: int
    members_by_status: dict
    members_by_engagement: dict
    members_by_location: dict
    members_by_country: dict
    team_sizes: dict

File: apps/dashboard/src/features/team-management/schemas.ts
import {
  z,
} from "zod";

import {
  engagementTypes,
  employmentStatuses,
  teamTypes,
  workLocationTypes,
} from "./types";

const metadataSchema =
  z.record(
    z.string(),
    z.unknown(),
  );

const countMapSchema =
  z.record(
    z.string(),
    z.number().int().nonnegative(),
  );

export const teamSchema =
  z.object({
    id: z.string().uuid(),
    name: z.string(),
    slug: z.string(),
    team_type: z.enum(teamTypes),
    description: z.string(),
    parent_id:
      z.string().uuid().nullable(),
    parent_name:
      z.string().nullable(),
    manager_id:
      z.string().uuid().nullable(),
    manager_name:
      z.string().nullable(),
    is_active: z.boolean(),
    is_public: z.boolean(),
    sort_order:
      z.number().int().nonnegative(),
    metadata: metadataSchema,
    member_count:
      z.number().int().nonnegative(),
  });

export const teamMembershipSchema =
  z.object({
    id: z.string().uuid(),
    team_id: z.string().uuid(),
    team_name: z.string(),
    role_title: z.string(),
    is_primary: z.boolean(),
    is_active: z.boolean(),
    joined_at: z.string(),
    left_at: z.string().nullable(),
    sort_order:
      z.number().int().nonnegative(),
  });

export const teamMemberServiceSchema =
  z.object({
    id: z.string().uuid(),
    service_id: z.string().uuid(),
    service_title: z.string(),
    expertise_level: z.string(),
    years_of_experience:
      z.union([
        z.number(),
        z.string(),
      ]).nullable(),
    is_primary: z.boolean(),
    is_public: z.boolean(),
    sort_order:
      z.number().int().nonnegative(),
  });

export const teamMemberSchema =
  z.object({
    id: z.string().uuid(),
    user_id:
      z.string().uuid().nullable(),
    employee_code: z.string(),
    first_name: z.string(),
    last_name: z.string(),
    full_name: z.string(),
    preferred_name: z.string(),
    display_name: z.string(),
    job_title: z.string(),
    professional_title: z.string(),
    email: z.string(),
    phone: z.string(),
    public_email: z.string(),
    public_phone: z.string(),
    profile_image_id:
      z.string().uuid().nullable(),
    bio: z.string(),
    short_bio: z.string(),
    qualifications: z.string(),
    years_of_experience:
      z.number().int().nonnegative().nullable(),
    engagement_type:
      z.enum(engagementTypes),
    employment_status:
      z.enum(employmentStatuses),
    work_location_type:
      z.enum(workLocationTypes),
    office_location: z.string(),
    country: z.string(),
    timezone_name: z.string(),
    joined_at: z.string().nullable(),
    employment_ended_at:
      z.string().nullable(),
    reports_to_id:
      z.string().uuid().nullable(),
    reports_to_name:
      z.string().nullable(),
    linkedin_url: z.string(),
    github_url: z.string(),
    portfolio_url: z.string(),
    website_url: z.string(),
    is_leadership: z.boolean(),
    is_public: z.boolean(),
    is_featured: z.boolean(),
    is_current: z.boolean(),
    sort_order:
      z.number().int().nonnegative(),
    metadata: metadataSchema,
    memberships:
      z.array(teamMembershipSchema),
    services:
      z.array(teamMemberServiceSchema),
  });

export const teamListSchema =
  z.array(teamSchema);

export const teamMemberListSchema =
  z.array(teamMemberSchema);

export const teamManagementDashboardSchema =
  z.object({
    total_teams:
      z.number().int().nonnegative(),
    active_teams:
      z.number().int().nonnegative(),
    public_teams:
      z.number().int().nonnegative(),
    total_members:
      z.number().int().nonnegative(),
    active_members:
      z.number().int().nonnegative(),
    members_on_leave:
      z.number().int().nonnegative(),
    inactive_members:
      z.number().int().nonnegative(),
    public_members:
      z.number().int().nonnegative(),
    featured_members:
      z.number().int().nonnegative(),
    leadership_members:
      z.number().int().nonnegative(),
    members_without_primary_team:
      z.number().int().nonnegative(),
    members_without_manager:
      z.number().int().nonnegative(),
    members_by_status:
      countMapSchema,
    members_by_engagement:
      countMapSchema,
    members_by_location:
      countMapSchema,
    members_by_country:
      countMapSchema,
    team_sizes:
      countMapSchema,
  });
